### PR TITLE
fix(csv): strip UTF-8 BOM and drop trailing empty header column in CsvToIon

### LINKED ISSUE #17646: Error on random quoted fields name with serdes CsvToIon
### Describe the issue

CsvToIon adds quotes around certain field names in the Ion file, causing downstream errors :

source example (CSV):
```
code_insee;nom_commune;code_departement;dose_rayonnements_telluriques;dose_rayonnements_cosmiques;dose_radon_maison_individuelle;dose_radon_habitat_collectif;dose_depots_essais_atmospheriques_et_tchernobyl
1001;L'Abergement-Clémenciat;1;636;323;4558;2482;9
1002;L'Abergement-de-Varey;1;369;335;3695;2189;9
1004;Ambérieu-en-Bugey;1;486;325;3029;1963;9
...
```

ion file (with quoted field on the first causing error with IonToParquet downstream):
```
{'﻿code_insee':"1001",nom_commune:"L'Abergement-Clémenciat",code_departement:"1",dose_rayonnements_telluriques:"636",dose_rayonnements_cosmiques:"323",dose_radon_maison_individuelle:"4558",dose_radon_habitat_collectif:"2482",dose_depots_essais_atmospheriques_et_tchernobyl:"9",'':""}
{'﻿code_insee':"1002",nom_commune:"L'Abergement-de-Varey",code_departement:"1",dose_rayonnements_telluriques:"369",dose_rayonnements_cosmiques:"335",dose_radon_maison_individuelle:"3695",dose_radon_habitat_collectif:"2189",dose_depots_essais_atmospheriques_et_tchernobyl:"9",'':""}
{'﻿code_insee':"1004",nom_commune:"Ambérieu-en-Bugey",code_departement:"1",dose_rayonnements_telluriques:"486",dose_rayonnements_cosmiques:"325",dose_radon_maison_individuelle:"3029",dose_radon_habitat_collectif:"1963",dose_depots_essais_atmospheriques_et_tchernobyl:"9",'':""}
...
```

### Environment

- Kestra Version: 1.3.29

### PR DESCRIPTION
**Pull request by Lalithkishore1224**
## What
Fixes two bugs in `CsvToIon` reported in kestra-io/kestra#17646.

## Why
1. A UTF-8 BOM at the start of a CSV file was leaking into the first header's
   name, causing only that field to be quoted (inconsistently with every
   other field) in the resulting Ion output.
2. A trailing field separator on the header line produced a nameless empty
   field on every row, which broke downstream `IonToParquet` conversion
   (Parquet requires real column names).

## Before / After
Reproduced with the exact sample data from the issue (French INSEE radiation
dataset, `;`-separated, UTF-8 BOM):

**Before:**
{'﻿code_insee':"1001",nom_commune:"L'Abergement-Clémenciat",code_departement:"1",...}

**After:**
{code_insee:"1001",nom_commune:"L'Abergement-Clémenciat",code_departement:"1",...}

## How
- Strip a leading UTF-8 BOM using Apache commons-io `BOMInputStream` before the
  CSV parser reads the stream. commons-io is declared as an explicit dependency,
  version managed by the `io.kestra:platform` BOM.
- Handle empty-named header columns via a new `onEmptyHeader` property:
  - `DROP` (default): drop trailing unnamed columns and log a warning. If the
    whole header is empty, keep every column instead of collapsing rows to `{}`.
  - `RENAME`: keep every column and name unnamed ones `col_0`, `col_1`, ...,
    disambiguated so a generated name never overwrites a real column that already
    has that name. No data is lost.
- Row field counts are still validated against the raw header count, so
  `onBadLines` detection is unaffected. An empty header name in the middle of the
  header is preserved as-is (structural data, not a trailing-separator artifact).

## Testing
`CsvToIonWriterTest` covers: BOM stripping, BOM plus a trailing separator together,
single and multiple trailing empty columns, an all-empty header, a middle empty
header, `DROP` vs `RENAME` on data-bearing trailing columns, the `RENAME`
name-collision case, and a row missing the header's trailing separator being
flagged as a bad line.

Full `CsvToIonWriterTest` suite: 20 passing, 0 failing.

Fixes kestra-io/kestra#17646

### CODE REVIEW

File: src/main/java/io/kestra/plugin/serdes/csv/CsvToIon.java [hunk 1]
```diff
@@ -298,4 +308,25 @@ private CsvReader<CsvRecord> csvReader(Reader reader, RunContext runContext) thr
         var handler = handlerBuilder.build();
         return builder.build(handler, reader);
     }
+
+    /**
+     * Strips a leading UTF-8 byte-order mark (EF BB BF), if present, so it doesn't leak into the
+     * first field's name/value (e.g. as an invisible character prepended to the first header).
+     */
+    private InputStream stripUtf8Bom(InputStream inputStream) throws IOException {
```
Comment (Malaydewangan09) [inline src/main/java/io/kestra/plugin/serdes/csv/CsvToIon.java]: I don't think we need the custom helper here. fastcsv already handles this: turn on `detectBomHeader(true)` and build from the InputStream with the charset, and it strips UTF-8/16/32 BOMs for us. That lets us drop this method and the PushbackInputStream.

File: src/main/java/io/kestra/plugin/serdes/csv/CsvToIon.java [hunk 2]
```diff
@@ -192,6 +193,15 @@ public Output run(RunContext runContext) throws Exception {
                         for (int i = 0; i < csvRecord.getFieldCount(); i++) {
                             headers.put(i, csvRecord.getField(i));
                         }
+                        rawHeaderFieldCount.set(csvRecord.getFieldCount());
+                        // Drop trailing empty-named columns (e.g. from a trailing field separator in the
+                        // header line): an unnamed field otherwise surfaces as an unusable "" field in the
+                        // Ion output and breaks downstream conversions (e.g. to Parquet).
+                        int trailing = csvRecord.getFieldCount();
+                        while (trailing > 0 && headers.get(trailing - 1) != null && headers.get(trailing - 1).isEmpty()) {
```
Comment (Malaydewangan09) [inline src/main/java/io/kestra/plugin/serdes/csv/CsvToIon.java]: One thing to watch here: this quietly removes the column and its data, and only when the empty header is at the end. If a header is empty in the middle it still slips through as an empty field name. Could we log when we drop a column, or give it a name instead, and handle unnamed columns in general rather than just trailing ones?

File: src/main/java/io/kestra/plugin/serdes/csv/CsvToIon.java [hunk 3]
```diff
@@ -211,9 +221,9 @@ public Output run(RunContext runContext) throws Exception {
                             }
                             return Mono.empty();
                         }
-                        if (r.getFieldCount() != headers.size()) {
+                        if (r.getFieldCount() != rawHeaderFieldCount.get()) {
```
Comment (Malaydewangan09) [inline src/main/java/io/kestra/plugin/serdes/csv/CsvToIon.java]: This leaves us with two counts to keep in sync: the rows are checked against `rawHeaderFieldCount` (the raw count) but the output loops over the trimmed `headers`. It works, but it's easy to break later. Might be cleaner to work out the effective columns once and use that everywhere.

File: src/test/java/io/kestra/plugin/serdes/csv/CsvToIonWriterTest.java [hunk 4]
```diff
@@ -194,6 +196,77 @@ void largeQuotedFieldParsesWhenBufferSufficient() throws Exception {
...
+
+        assertThat(rows, hasSize(1));
+        @SuppressWarnings("unchecked")
+        Map<String, Object> row = (Map<String, Object>) rows.get(0);
+        assertThat(row.keySet(), hasItem("code_insee"));
+        assertThat(row.get("code_insee"), is("1001"));
+    }
+
+    @Test
+    void trailingEmptyHeaderColumnIsDropped() throws Exception {
```
Comment (Malaydewangan09) [inline src/test/java/io/kestra/plugin/serdes/csv/CsvToIonWriterTest.java:276]: Worth adding a case with a genuinely empty-named column that carries data, to lock the intended drop behavior.

File: src/main/java/io/kestra/plugin/serdes/csv/CsvToIon.java (hunk 1 above)
Comment (Lalithkishore1224) [reply src/main/java/io/kestra/plugin/serdes/csv/CsvToIon.java]: detectBomHeader(true) instead of the custom helper: That was actually my first approach too, but it introduced a regression — enabling it requires building from an InputStream+Charset (not a Reader), and that path shrinks the effective capacity of a user-configured maxBufferSize enough that the existing largeQuotedFieldParsesWhenBufferSufficient test (128-byte buffer, 50-char quoted field) started failing. Switched to peeking 3 raw bytes via PushbackInputStream instead — fully decoupled from fastcsv's internal buffering — and documented the tradeoff in a Javadoc comment so it's not re-litigated blind later.

File: src/main/java/io/kestra/plugin/serdes/csv/CsvToIon.java (hunk 2 above)
Comment (Lalithkishore1224) [reply src/main/java/io/kestra/plugin/serdes/csv/CsvToIon.java]: Log when a column is dropped: Added a warning log naming how many trailing unnamed columns were dropped and their indices. Chose drop+log over auto-naming, since synthesizing a name risks masking a real data-quality issue as if it were intentional.

File: src/main/java/io/kestra/plugin/serdes/csv/CsvToIon.java (hunk 3 above)
Comment (Lalithkishore1224) [reply src/main/java/io/kestra/plugin/serdes/csv/CsvToIon.java]: Two counts to keep in sync: Refactored to a single source of truth — headers is no longer mutated and always reflects the raw field count (used for onBadLines validation, unchanged from before this PR). Trailing artifact columns are tracked separately as an immutable index set, computed once, consulted only at output time.

File: src/test/java/io/kestra/plugin/serdes/csv/CsvToIonWriterTest.java (hunk 4 above)
Comment (Lalithkishore1224) [reply src/test/java/io/kestra/plugin/serdes/csv/CsvToIonWriterTest.java:276]: Test for empty-named column with data: Added emptyHeaderNameInTheMiddleIsPreservedNotDropped, locking in that only trailing empty names are treated as the artifact.

**Comment by Malaydewangan09** [comment]
@Lalithkishore1224, If you've addressed any of the requested changes from the PR review, , make sure to click the Resolve conversation button for that particular change to signal to the reviewer that you've completed what was requested.

File: src/main/java/io/kestra/plugin/serdes/csv/CsvToIon.java [hunk 5]
```diff
@@ -298,4 +318,31 @@ private CsvReader<CsvRecord> csvReader(Reader reader, RunContext runContext) thr
...
+     * This is done manually rather than via {@code CsvReaderBuilder#detectBomHeader(true)} because
+     * that option only takes effect when building from an {@code InputStream}, and doing so shrinks
+     * the effective capacity of a user-configured {@code maxBufferSize} enough to make small buffer
+     * sizes (e.g. the ones used to bound pathological fields) fail unexpectedly. Peeking 3 raw bytes
+     * here is fully decoupled from fastcsv's internal buffering.
+     */
+    private InputStream stripUtf8Bom(InputStream inputStream) throws IOException {
+        PushbackInputStream pushbackInputStream = new PushbackInputStream(inputStream, 3);
+        byte[] possibleBom = new byte[3];
+        int read = pushbackInputStream.read(possibleBom, 0, 3);
```
Comment (Malaydewangan09) [inline src/main/java/io/kestra/plugin/serdes/csv/CsvToIon.java]: `read(possibleBom, 0, 3)` can return fewer than 3 bytes. Fine on local storage (`BufferedInputStream`), but S3/GCS return an unbuffered stream where a short read is legal, so the BOM leaks through. Simpler to drop the helper and use commons-io `BOMInputStream`, already on the classpath:

```java
BOMInputStream.builder()
    .setInputStream(runContext.storage().getFile(rFrom))
    .setByteOrderMarks(ByteOrderMark.UTF_8)
    .get()
```

Strips the UTF-8 BOM, reads correctly, removes ~20 lines and the bug.

File: src/main/java/io/kestra/plugin/serdes/csv/CsvToIon.java (hunk 2 above)
Comment (Malaydewangan09) [reply src/main/java/io/kestra/plugin/serdes/csv/CsvToIon.java]: Fine to keep a real middle empty name. Just flagging that a middle one still comes out as a `""` key and breaks the same Parquet conversion, so this is trailing-only by design, not a full fix. Two empty middle columns would also collapse into one `""` key (pre-existing).

File: src/main/java/io/kestra/plugin/serdes/csv/CsvToIon.java [hunk 6]
```diff
@@ -171,6 +171,7 @@ public Output run(RunContext runContext) throws Exception {
             var rHeaderValue = runContext.render(header).as(Boolean.class).orElseThrow();
             var rSkipRowsValue = runContext.render(this.skipRows).as(Integer.class).orElseThrow();
             Map<Integer, String> headers = new TreeMap<>();
+            Set<Integer> trailingEmptyHeaderColumns = new HashSet<>();
```
Comment (Malaydewangan09) [inline src/main/java/io/kestra/plugin/serdes/csv/CsvToIon.java]: Trailing empties are always a contiguous run, so this could be one `int effectiveHeaderCount` instead of a `Set` plus a per-row `.contains()`. Same behaviour, less machinery. Also `headers.get(trailing - 1) != null` can't fire (fastcsv returns `""`, not null). Non-blocking.

**Comment by Lalithkishore1224** [comment]
Thanks

BOM short-read: You're right — read(buf, 0, 3) isn't guaranteed to fill the buffer even with more data available, only EOF gives that guarantee, so a short read from a streamed backend (S3/GCS) could let the BOM through. Switched to BOMInputStream as you suggested — added commons-io as an explicit dependency in build.gradle rather than leaning on the transitive resolution.

Set + contains() → single count: Agreed, done. Replaced with effectiveHeaderCount and dropped the dead null check too (fastcsv returns "" for an empty field, never null, so that branch could never fire).

Full suite re-run: 333 passing, 0 failing.

File: src/main/java/io/kestra/plugin/serdes/csv/CsvToIon.java [hunk 7]
```diff
@@ -192,6 +198,23 @@ public Output run(RunContext runContext) throws Exception {
                         for (int i = 0; i < csvRecord.getFieldCount(); i++) {
                             headers.put(i, csvRecord.getField(i));
                         }
+                        // Trailing empty-named columns are always a contiguous run at the end of the
+                        // header line (e.g. from a trailing field separator): left as-is, an unnamed
+                        // field surfaces as an unusable "" field in the Ion output and breaks downstream
+                        // conversions (e.g. to Parquet). An empty name earlier in the header is a
+                        // structural property of the data itself and is kept as-is.
+                        int trailing = csvRecord.getFieldCount();
+                        while (trailing > 0 && headers.get(trailing - 1).isEmpty()) {
```
Comment (Malaydewangan09) [inline src/main/java/io/kestra/plugin/serdes/csv/CsvToIon.java]: Edge case: if the whole header is empty (like `;;;`), this trims to 0 columns and every row comes out as an empty `{}`, with only a warning. Can we trim only when at least one real header is left? A test for an all-empty header would help.

**Comment by Malaydewangan09** [comment]
Two housekeeping notes, not code:

- The task now auto-strips a BOM and auto-drops trailing empty columns, but the `@Schema` description and the plugin doc don't mention it. A one-line note would give someone a trail if a column goes missing.
- The PR description still says the BOM is stripped with a "3-byte PushbackInputStream peek" - the final code uses commons-io `BOMInputStream`. Worth updating so the record matches.

One more (code, but on an unchanged line so no inline): the row check uses the raw header count, so a row that leaves off the header's trailing `;` gets flagged as a bad line. Fine for this issue, worth a test to show it's intended.

File: src/main/java/io/kestra/plugin/serdes/csv/CsvToIon.java [hunk 8]
```diff
@@ -160,7 +162,10 @@ public Output run(RunContext runContext) throws Exception {
         try (
             Reader reader = new BufferedReader(
                 new InputStreamReader(
-                    runContext.storage().getFile(rFrom),
+                    BOMInputStream.builder()
+                        .setInputStream(runContext.storage().getFile(rFrom))
+                        .setByteOrderMarks(ByteOrderMark.UTF_8)
```
Comment (Malaydewangan09) [inline src/main/java/io/kestra/plugin/serdes/csv/CsvToIon.java]: Small note: this only strips a UTF-8 BOM. Fine for this issue, just won't cover UTF-16/32 BOMs.

File: src/test/java/io/kestra/plugin/serdes/csv/CsvToIonWriterTest.java [hunk 9]
```diff
@@ -194,6 +196,111 @@ void largeQuotedFieldParsesWhenBufferSufficient() throws Exception {
...
```
Comment (Malaydewangan09) [inline src/test/java/io/kestra/plugin/serdes/csv/CsvToIonWriterTest.java:276]: Could add one more case: two or more trailing empty columns (`a;b;;`), to confirm the whole run gets trimmed.

**Comment by Lalithkishore1224** [comment]
Thanks — addressed all of these:

All-empty header edge case: Good catch. Added a guard: if trimming trailing empty columns would leave zero columns (i.e. the whole header is empty), we skip the trim entirely and keep every column instead of collapsing every row to {}. Added allEmptyHeaderIsNotTrimmedToZeroColumns to lock this in.

Docs: Added a line to the @Schema description noting the BOM-strip and trailing-empty-column-drop behavior, so it's discoverable from the plugin doc, not just the changelog.

PR description mismatch: Fixed — updated to say BOMInputStream (commons-io), matching the actual code. Thanks for catching that.

Row omitting the header's trailing separator: Confirmed intended — added rowMissingHeadersTrailingSeparatorIsFlaggedAsBadLine, asserting it raises the field-count-mismatch error rather than silently passing.

UTF-8-only BOM: Deliberate scope decision, not an oversight — this PR only handles the UTF-8 case reported in the issue (also Kestra's default charset). Happy to extend setByteOrderMarks(...) to cover UTF-16/32 in a follow-up if there's an actual report of it, but didn't want to expand scope speculatively here.

Multiple trailing empty columns test: Added multipleTrailingEmptyHeaderColumnsAreDropped (a;b;;), confirming the whole contiguous run gets dropped.

Full suite: 336 passing, 0 failing.

**Comment by Malaydewangan09** [comment]
Pushed a small follow-up commit (`391656b`) after another review pass. In plain terms:

1. **No more silent data loss.** If two header columns end up with the same name (blank names, or an all-blank header like `;;;`), only one of them used to survive and nothing told you. It now logs a clear warning so you know a column got dropped, and suggests `RENAME` if you want to keep everything.
2. **BOM stripping is safer.** The auto-remove of a UTF-8 marker at the start of a file now only happens when the file is actually read as UTF-8. For other encodings those bytes are real text, so we leave them alone.
3. **Docs formatting.** Switched the `onEmptyHeader` description to the standard multi-line format used elsewhere, and removed a `@Schema` annotation on the enum that did nothing.
4. **New test** confirming those start-of-file bytes are kept when a non-UTF-8 charset is used.

One thing left as-is on purpose: a data row that leaves off the header's trailing separator is still flagged as a bad line. That is the intended, already-tested behavior and matches the reported case, so I did not change it.

All `CsvToIonWriterTest` tests pass (21).
